Idrall/Check.idr

```diff
@@ -399,8 +449,27 @@ mutual
   readBackTyped ctx (VOptional ty) (VSome a) = do
     a' <- readBackTyped ctx ty a
     Right (ESome a')
+  readBackTyped ctx (VConst _) (VUnion a) = do
+    a' <- traverse (readBackUnion ctx) (toList a)
+    Right (EUnion (fromList a'))
+  readBackTyped ctx (VUnion _) (VInject a k arg) =
+    let kV = lookup k a in
+    case (kV, arg) of
+         (Just (Just k'), Just arg') =>
+           do aRB <- traverse (readBackUnion ctx) (toList a)
+              arg'' <- readBackTyped ctx k' arg'
+              Right (EApp (EField (EUnion (fromList aRB)) k) (arg''))
+         (Just Nothing, Just arg') => Left (FieldArgMismatchError ("No type for param " ++ show k))
+         (Just _, _) => Right (EField
+                                      (EUnion (fromList !(traverse (readBackUnion ctx) (toList a)))) k)
+         (Nothing, _) => Left (FieldNotFoundError (show k))
   readBackTyped _ t v = Left (ReadBackError ("error reading back: " ++ (show v) ++ " of type: " ++ (show t)))
 
+  readBackUnion : Ctx -> (FieldName, Maybe Value) -> Either Error (FieldName, Maybe (Expr Void))
+  readBackUnion ctx (k, Nothing) = Right (k, Nothing)
+  readBackUnion ctx (k, Just v) = Right (k, Just !(readBackTyped ctx (VConst CType) v))
+    -- TODO is (VConst CType) always right here ^^^? Looks like rBT ignores the Ty param when reading back VConsts so maybe?
```

Comment:
I think it is fine, since you are only using it for the type

Reply:
Aye, was thinking so too. Andras's branch doesn't have a type param at all for reading back, not sure the implications of that, so I guess when reading back the straight forward type values (`VType`, `VKind`, `VBool`) should be safe to ignore that type param.